pkg/chunk/purger/purger.go

```diff
@@ -0,0 +1,566 @@
+	planNo          int
+	userID          string
+	deleteRequestID string
+	logger          log.Logger
+}
+
+type workerJobExecutionStatus struct {
+	workerJob
+	err error
+}
+
+// DataPurger does the purging of data which is requested to be deleted
+type DataPurger struct {
+	cfg           DataPurgerConfig
+	deleteStore   *chunk.DeleteStore
+	chunkStore    chunk.Store
+	storageClient chunk.ObjectClient
+
+	executePlansChan chan deleteRequestWithLogger
+	workerJobChan    chan workerJob
+
+	// we would only allow processing of singe delete request at a time since delete requests touching same chunks could change the chunk IDs of partially deleted chunks
+	// and break the purge plan for other requests
+	inProcessRequestIDs    map[string]string
+	inProcessRequestIDsMtx sync.RWMutex
+
+	pendingPlansCount    map[string]int // per request pending plan count
+	pendingPlansCountMtx sync.Mutex
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewDataPurger creates a new DataPurger
+func NewDataPurger(cfg DataPurgerConfig, deleteStore *chunk.DeleteStore, chunkStore chunk.Store, storageClient chunk.ObjectClient) (*DataPurger, error) {
+	dataPurger := DataPurger{
+		cfg:                 cfg,
+		deleteStore:         deleteStore,
+		chunkStore:          chunkStore,
+		storageClient:       storageClient,
+		executePlansChan:    make(chan deleteRequestWithLogger, 50),
+		workerJobChan:       make(chan workerJob, 50),
+		inProcessRequestIDs: map[string]string{},
+		pendingPlansCount:   map[string]int{},
+		quit:                make(chan struct{}),
+	}
+
+	dataPurger.runWorkers()
+	go dataPurger.jobScheduler()
+
+	err := dataPurger.loadInprocessDeleteRequests()
+	if err != nil {
+		return nil, err
+	}
+
+	return &dataPurger, nil
+}
+
+// Run keeps pulling delete requests for planning
+func (dp *DataPurger) Run() {
+	dp.wg.Add(1)
+	defer dp.wg.Done()
+
+	pullDeleteRequestsToPlanDeletesTicker := time.NewTicker(time.Hour)
+	defer pullDeleteRequestsToPlanDeletesTicker.Stop()
+
+	for {
+		select {
+		case <-pullDeleteRequestsToPlanDeletesTicker.C:
+			err := dp.pullDeleteRequestsToPlanDeletes()
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "error pulling delete requests for building plans", "err", err)
+			}
+		case <-dp.quit:
+			return
+		}
+	}
+}
+
+// Stop stops all background workers/loops
+func (dp *DataPurger) Stop() {
+	close(dp.quit)
+	dp.wg.Wait()
+}
+
+func (dp *DataPurger) processWorkerJobExecutionStatus(status workerJobExecutionStatus) {
+	if status.err != nil {
+		level.Error(status.logger).Log("msg", "error executing delete plan",
+			"plan_no", status.planNo, "err", status.err)
+		return
+	}
+
+	err := dp.removeDeletePlan(context.Background(), status.userID, status.deleteRequestID, status.planNo)
+	if err != nil {
+		level.Error(status.logger).Log("msg", "error removing delete plan",
+			"plan_no", status.planNo, "err", err)
+	} else {
+		dp.pendingPlansCountMtx.Lock()
+		dp.pendingPlansCount[status.deleteRequestID]--
+
```

Comment:
Use `defer dp.pendingPlansCountMtx.Unlock()` here?

Reply:
Actually I wanted to do that but I am working on 2 mutexes here so I want to avoid taking lock on both of them at a time. While its not a problem to take lock on both of them now but future changes might cause a deadlock if order of locks change in some other piece of code.